core/unit_test/tools/include/ToolTestingUtilities.hpp

```diff
@@ -0,0 +1,256 @@
+ * classes are empty
+ */
+struct BeginParallelForEvent : public BeginOperation<BeginParallelForEvent> {
+  BeginParallelForEvent(
+      std::string n,
+      const uint32_t devID = EventBase::unspecified_sentinel<uint32_t>,
+      uint64_t k           = EventBase::unspecified_sentinel<uint64_t>)
+      : BeginOperation<BeginParallelForEvent>(n, devID, k) {}
+  virtual ~BeginParallelForEvent() = default;
+};
+struct BeginParallelReduceEvent
+    : public BeginOperation<BeginParallelReduceEvent> {
+  BeginParallelReduceEvent(
+      std::string n,
+      const uint32_t devID = EventBase::unspecified_sentinel<uint32_t>,
+      uint64_t k           = EventBase::unspecified_sentinel<uint64_t>)
+      : BeginOperation<BeginParallelReduceEvent>(n, devID, k) {}
+  virtual ~BeginParallelReduceEvent() = default;
+};
+struct BeginParallelScanEvent : public BeginOperation<BeginParallelScanEvent> {
+  BeginParallelScanEvent(
+      std::string n,
+      const uint32_t devID = EventBase::unspecified_sentinel<uint32_t>,
+      uint64_t k           = EventBase::unspecified_sentinel<uint64_t>)
+      : BeginOperation<BeginParallelScanEvent>(n, devID, k) {}
+  virtual ~BeginParallelScanEvent() = default;
+};
+struct BeginFenceEvent : public BeginOperation<BeginFenceEvent> {
+  BeginFenceEvent(
+      std::string n,
+      const uint32_t devID = EventBase::unspecified_sentinel<uint32_t>,
+      uint64_t k           = EventBase::unspecified_sentinel<uint64_t>)
+      : BeginOperation<BeginFenceEvent>(n, devID, k) {}
+  virtual ~BeginFenceEvent() = default;
+};
+
+struct EndParallelForEvent : public EndOperation<EndParallelForEvent> {
+  EndParallelForEvent(uint64_t k = EventBase::unspecified_sentinel<uint64_t>)
+      : EndOperation<EndParallelForEvent>(k) {}
+  virtual ~EndParallelForEvent() = default;
+};
+struct EndParallelReduceEvent : public EndOperation<EndParallelReduceEvent> {
+  EndParallelReduceEvent(uint64_t k = EventBase::unspecified_sentinel<uint64_t>)
+      : EndOperation<EndParallelReduceEvent>(k) {}
+  virtual ~EndParallelReduceEvent() = default;
+};
+struct EndParallelScanEvent : public EndOperation<EndParallelScanEvent> {
+  EndParallelScanEvent(uint64_t k = EventBase::unspecified_sentinel<uint64_t>)
+      : EndOperation<EndParallelScanEvent>(k) {}
+  virtual ~EndParallelScanEvent() = default;
+};
+struct EndFenceEvent : public EndOperation<EndFenceEvent> {
+  EndFenceEvent(uint64_t k = EventBase::unspecified_sentinel<uint64_t>)
+      : EndOperation<EndFenceEvent>(k) {}
+  virtual ~EndFenceEvent() = default;
+};
+
+bool compare_event_vectors(const std::vector<EventBasePtr>& expected,
+                           const std::vector<EventBasePtr>& found) {
+  auto expected_size = expected.size();
+  if (found.size() != expected_size) {
+    std::cout << "Expected " << expected_size << " events, got " << found.size()
+              << std::endl;
+    for (auto entry : found) {
+      std::cout << "Entry: " << entry->repr() << std::endl;
+    }
```

Comment:
I wouldn't print anything here if sizes do not match.  This makes me wish we were using gmock as well since it comes with a facility for comparing collections of elements.

Reply:
I disagree on this one. I actually had a bug in a test where I had six events when I expected two, and didn't know why. Printing out what events were actually found helped me see "oh, I forgot about the fence events from the parallel_for." Not something I feel _very_ strongly about, but I do see value in this debug output

Reply:
Outdated because the new code has a different error output mechanism. Can reopen a new discussion on the new output mechanisms